### PR TITLE
Pull iso as a container

### LINKED ISSUE #396: Pull base ISO as containers instead of http download
Currently the build ISO feature downloads the ISO from an http url within the build environment. That means multiple builds based on the same ISO results into multiple downloads of the same exact bits. In order to improve this I'd suggest to pull the ISO image as a container, as an init continer of the build pod.

Why?

* We can reuse the same pulled image for many machine registrations
* We can easily relay on the same distribution scheme as we have for OS containers (registry and channel concept)
* We relay on kubernetes garbage collector for the cleanup of unused ISOs
* Nothing needs to be downloaded for an ISO rebuild over the same node.

What needs to be changed?

* Adapt current build-iso to include labels and tags as expected and be perfectly aligned with node-image
* Include busybox or similar in containerzied ISO to copy contents to somewhere else
* Adapt elemental operator to pull image and copy container contents instead of downloading from the network
* Include ISO field in ManagedOSVersions to link ISOs with released image?

### PR DESCRIPTION
**Pull request by davidcassany**
This PR allows to pull the base ISO as a container or from an HTTP URL. The logic to use one or the other is based done by parsing the URL, any HTTP URL will use curl to download the ISO and anything else will attempted to be used as a container reference.

Fixes #396

### CODE REVIEW

File: controllers/seedimage_controller.go [hunk 1]
```diff
@@ -448,12 +448,6 @@ func (r *SeedImageReconciler) getRancherServerAddress(ctx context.Context) (stri
 
 func fillBuildImagePod(name, namespace, buildImg, baseImg, configMap string, pullPolicy corev1.PullPolicy, deadline int32) *corev1.Pod {
 	const volLim = 4 * 1024 * 1024 * 1024 // 4 GiB
-	const volRes = 2 * 1024 * 1024 * 1024 // 2 GiB
```
Comment (davidcassany) [inline controllers/seedimage_controller.go:451]: @fgiudici I am not sure we actually need that, I'd say copying everything directly to the shared volume is just enough.

I faced this because if pulling the ISO from a container I need to share it to the builder container and the obvious way to do so is simply reusing the one we already have to store the built ISO.

File: controllers/seedimage_controller.go [hunk 2]
```diff
@@ -539,6 +508,57 @@ func fillBuildImagePod(name, namespace, buildImg, baseImg, configMap string, pul
...
+		buildCommands = append([]string{fmt.Sprintf("curl -Lo %s %s", baseIsoPath, baseImg)}, buildCommands...)
+	} else {
+		// If baseImg is not an HTTP url assume it is an image reference
+		containers = append(
+			containers, corev1.Container{
+				Name:            "baseiso",
+				Image:           baseImg,
+				ImagePullPolicy: corev1.PullIfNotPresent,
+				Command:         []string{"busybox", "sh", "-c"},
+				Args:            []string{fmt.Sprintf("busybox cp /elemental-iso/*.iso %s", baseIsoPath)},
```
Comment (davidcassany) [inline controllers/seedimage_controller.go:529]: The base iso is copied to the shared volume.

File: controllers/seedimage_controller.go [hunk 3]
```diff
@@ -539,6 +508,57 @@ func fillBuildImagePod(name, namespace, buildImg, baseImg, configMap string, pul
 	return pod
 }
 
+func defineInitContainers(baseImg, buildImg string, pullPolicy corev1.PullPolicy) []corev1.Container {
+	const baseIsoPath = "/iso/base.iso"
+
+	containers := []corev1.Container{}
+	buildCommands := []string{
+		fmt.Sprintf("xorriso -indev %s -outdev /iso/elemental.iso -map /overlay / -boot_image any replay", baseIsoPath),
```
Comment (davidcassany) [inline controllers/seedimage_controller.go:516]: The iso is downloaded to the shared volume, so no resources additional resources are required

File: controllers/seedimage_controller.go [hunk 4]
```diff
@@ -448,12 +448,6 @@ func (r *SeedImageReconciler) getRancherServerAddress(ctx context.Context) (stri
 
 func fillBuildImagePod(name, namespace, buildImg, baseImg, configMap string, pullPolicy corev1.PullPolicy, deadline int32) *corev1.Pod {
 	const volLim = 4 * 1024 * 1024 * 1024 // 4 GiB
```
Comment (davidcassany) [inline controllers/seedimage_controller.go:450]: We could increase it to 6GiB (the bare minimum required with the previous setup) or 8GiB (to match the maximum size allowed with the previous setup)

File: controllers/seedimage_controller.go [hunk 5]
```diff
@@ -539,6 +508,57 @@ func fillBuildImagePod(name, namespace, buildImg, baseImg, configMap string, pul
...
+
+	containers := []corev1.Container{}
+	buildCommands := []string{
+		fmt.Sprintf("xorriso -indev %s -outdev /iso/elemental.iso -map /overlay / -boot_image any replay", baseIsoPath),
+	}
+
+	if util.IsHTTP(baseImg) {
+		buildCommands = append([]string{fmt.Sprintf("curl -Lo %s %s", baseIsoPath, baseImg)}, buildCommands...)
+	} else {
+		// If baseImg is not an HTTP url assume it is an image reference
```
Comment (davidcassany) [inline controllers/seedimage_controller.go:522]: Would it be worth having an image reference validation here? I am not sure, we have already other references in other places without any sort of validation. If we consider so I'd do it in a separate PR to try to validate any image reference provided by the user.

File: controllers/seedimage_controller.go (hunk 1 above)
Comment (fgiudici) [reply controllers/seedimage_controller.go:451]: well, this refers to the same volume, it is just the minimum disk space we require for the Pod to be created.
This is mainly for the kube API and the kubelet to ensure the destination node will have this minimum required disk space available.
If you drop it, it would default to 'volLim', which is the *maximum* space that can be used on the storage.
That means anyway, that for each seedimage pod running the kubelet will compute 'volLim' disk space as reserved and no more available on the node for other workloads.
I really think we should keep this, otherwise may affect pod allocation.

File: controllers/seedimage_controller.go (hunk 3 above)
Comment (fgiudici) [reply controllers/seedimage_controller.go:516]: We should clean it up then when done, otherwise we will occupy extra disk space on the node

File: controllers/seedimage_controller.go (hunk 5 above)
Comment (fgiudici) [reply controllers/seedimage_controller.go:522]: well, don't know... It would be good to have validation of the built ISO, that's a good point.
What about doing a check after the ISO is built, and before publishing it?
Anyway, sure, I would not add anything in this PR. We may track it in a new issue.

File: controllers/seedimage_controller.go (hunk 2 above)
Comment (fgiudici) [reply controllers/seedimage_controller.go:529]: pretty neat :-)

File: controllers/seedimage_controller.go [hunk 6]
```diff
@@ -465,32 +459,7 @@ func fillBuildImagePod(name, namespace, buildImg, baseImg, configMap string, pul
...
-					Name:            "build",
-					Image:           buildImg,
-					ImagePullPolicy: pullPolicy,
-					Resources: corev1.ResourceRequirements{
-						Limits: corev1.ResourceList{
-							corev1.ResourceEphemeralStorage: *resource.NewQuantity(volLim, resource.BinarySI),
-						},
-						Requests: corev1.ResourceList{
-							corev1.ResourceEphemeralStorage: *resource.NewQuantity(volRes, resource.BinarySI),
-						},
```
Comment (fgiudici) [inline controllers/seedimage_controller.go:479]: Well. you actually dropped all the ResourceRequirements: this here was just for scheduling, to ensure the destination node has enough disk space to be picked up to start the node (which actually will be the space reserved for the shared volume).

**Review by fgiudici — CHANGES_REQUESTED** [review]
Consuming the base ISO from a container image in this way is pretty neat! Like a lot the improvement!
I think anyway we should improve on the disk resource usage on the nodes. We should:
- leave the declaration of the Disk Resources we need to start the Pod, which will be of use to the kube API scheduling
- remove the base iso from the shared volume once we have built the final ISO: that will otherwise keep reserved some storage on the node

File: controllers/seedimage_controller.go (hunk 6 above)
Comment (davidcassany) [reply controllers/seedimage_controller.go:479]: I thought this was apart from the shared volume. I could be easily misunderstanding this feature ResourceRequirements here. I thought this ResourceRequirement here was only for the base image, so I assumed the Pod requirements are the shared volume plus this ResourceRequirement here. That's why I thought about downloading the base image to the shared volume, this way the resource constraints can be expressed in the same way regardless downloading it form HTTP or pulling it as a container, as in both cases the base image is copied/downloaded into the shared volume.

This is the reasoning of the change, but probably I am misunderstanding it.

Of course we adjust the share volume size to whatever we consider, to my understanding to be fully consistent with the former code it should be set to 8GiB (twice the max base image size).

File: controllers/seedimage_controller.go (hunk 3 above)
Comment (davidcassany) [reply controllers/seedimage_controller.go:516]: I was wondering about it too, but I could not see much benefit out of it, as up to my understanding the size reserved by the pod regardless being used or not. Hence to free resources the pods needs to be terminated. It's true however that after terminating the pod if the volume is smaller then is less likely to require k8s to start cleaning unused resources.

We can add a clean up call here, that's easy.

File: controllers/seedimage_controller.go (hunk 4 above)
Comment (davidcassany) [reply controllers/seedimage_controller.go:450]: Yes that is, so something like twice the max base image size. The thing is agreeing on what we consider the max, I am fine with 4Gib, so in that case we should set the volume to 8GiB.

File: controllers/seedimage_controller.go (hunk 5 above)
Comment (davidcassany) [reply controllers/seedimage_controller.go:522]: Not following you... There is no built done as this stage 😅 But yes, this sort of validations I'd move them to another issue as this is not exclusively related to this card.

**Comment by davidcassany** [comment]
@fgiudici I think now this is what we discussed. It always requires 4Gib to be present before starting the pod and the base image is deleted after the build.

File: controllers/seedimage_controller.go [hunk 7]
```diff
@@ -539,6 +508,68 @@ func fillBuildImagePod(name, namespace, buildImg, baseImg, configMap string, pul
...
+					MountPath: "/iso",
+				}, {
+					Name:      "config-map",
+					MountPath: "/overlay",
+				},
+			},
+		},
+	)
+
+	// Make sure the pod starts starts having enough disk for the maximum volume size
```
Comment (fgiudici) [inline controllers/seedimage_controller.go:560]:
```suggestion
	// Make sure the pod starts having enough disk for the maximum volume size
```

**Review by fgiudici — APPROVED** [review]
Great patch!

**Comment by davidcassany** [comment]
@fgiudici feel free to merge https://github.com/rancher/elemental-operator/pull/432 if you are done there. Then I'll rebase this one.
